src/source/Sdp/Serialize.c

```diff
@@ -1,47 +1,68 @@
-#define LOG_CLASS "SDP"
+#define LOG_CLASS "SDPSerialize"
 #include "../Include_i.h"
+#include "sdp_serializer.h"
 
-STATUS serializeVersion(UINT64 version, PCHAR* ppOutputData, PUINT32 pTotalWritten, PUINT32 pBufferSize)
+static STATUS serializeVersion(SdpSerializerContext_t* pCtx, UINT64 version)
 {
     ENTERS();
     STATUS retStatus = STATUS_SUCCESS;
-    UINT32 currentWriteSize = 0;
+    SdpResult_t sdpResult = SDP_RESULT_OK;
 
-    currentWriteSize = SNPRINTF(*ppOutputData, (*ppOutputData) == NULL ? 0 : *pBufferSize - *pTotalWritten,
-                                SDP_VERSION_MARKER "%" PRIu64 SDP_LINE_SEPARATOR, version);
+    sdpResult = SdpSerializer_AddU64(pCtx, SDP_TYPE_VERSION, version);
 
-    CHK(*ppOutputData == NULL || ((*pBufferSize - *pTotalWritten) >= currentWriteSize), STATUS_BUFFER_TOO_SMALL);
-    *pTotalWritten += currentWriteSize;
-    if (*ppOutputData != NULL) {
-        *ppOutputData += currentWriteSize;
+    if (sdpResult == SDP_RESULT_OK) {
+        retStatus = STATUS_SUCCESS;
+    } else if (sdpResult == SDP_RESULT_OUT_OF_MEMORY) {
+        retStatus = STATUS_BUFFER_TOO_SMALL;
+    } else {
+        retStatus = sdpResult;
     }
 
-CleanUp:
-
     LEAVES();
     return retStatus;
 }
 
-STATUS serializeOrigin(PSdpOrigin pSDPOrigin, PCHAR* ppOutputData, PUINT32 pTotalWritten, PUINT32 pBufferSize)
+static STATUS serializeOrigin(SdpSerializerContext_t* pCtx, PSdpOrigin pSDPOrigin)
 {
     ENTERS();
     STATUS retStatus = STATUS_SUCCESS;
-    UINT32 currentWriteSize = 0;
+    SdpResult_t sdpResult = SDP_RESULT_OK;
+    SdpOriginator_t origin;
 
     CHK(pSDPOrigin != NULL, STATUS_NULL_ARG);
 
     if (pSDPOrigin->userName[0] != '\0' && pSDPOrigin->sdpConnectionInformation.networkType[0] != '\0' &&
         pSDPOrigin->sdpConnectionInformation.addressType[0] != '\0' && pSDPOrigin->sdpConnectionInformation.connectionAddress[0] != '\0') {
-        currentWriteSize = SNPRINTF(*ppOutputData, (*ppOutputData) == NULL ? 0 : *pBufferSize - *pTotalWritten,
-                                    SDP_ORIGIN_MARKER "%s %" PRIu64 " %" PRIu64 " %s %s %s" SDP_LINE_SEPARATOR, pSDPOrigin->userName,
-                                    pSDPOrigin->sessionId, pSDPOrigin->sessionVersion, pSDPOrigin->sdpConnectionInformation.networkType,
-                                    pSDPOrigin->sdpConnectionInformation.addressType, pSDPOrigin->sdpConnectionInformation.connectionAddress);
-
-        CHK(*ppOutputData == NULL || ((*pBufferSize - *pTotalWritten) >= currentWriteSize), STATUS_BUFFER_TOO_SMALL);
-        *pTotalWritten += currentWriteSize;
-        if (*ppOutputData != NULL) {
-            *ppOutputData += currentWriteSize;
-        }
+        origin.pUserName = pSDPOrigin->userName;
+        origin.userNameLength = STRLEN(pSDPOrigin->userName);
```

Comment:
Let's always use `STRNLEN` instead of `STRLEN` for safety.

Reply:
Thanks for comment!
I replaced almost all STRLEN with STRNLEN. The only one left is `deserializeSessionDescription()` in Deserialize.c to calculate the length of sdpBytes, which we don't know the buffer size in this function.